clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/consumer/PlaintextConsumerFetchTest.java

```diff
@@ -0,0 +1,489 @@
+        Map<String, Object> config = Map.of(
+            GROUP_PROTOCOL_CONFIG, groupProtocol.name().toLowerCase(Locale.ROOT),
+            // one of the effects of this is that there will be some log reads where `0 > remaining limit bytes < message size`
+            // and we don't return the message because it's not the first message in the first non-empty partition of the fetch
+            // this behaves a little different from when remaining limit bytes is 0, and it's important to test it
+            FETCH_MAX_BYTES_CONFIG, 500,
+            MAX_PARTITION_FETCH_BYTES_CONFIG, 100,
+            // Avoid a rebalance while the records are being sent (the default is 6 seconds)
+            MAX_POLL_INTERVAL_MS_CONFIG, 20000
+        );
+        
+        try (Consumer<byte[], byte[]> consumer = cluster.consumer(config);
+             Producer<byte[], byte[]> producer = cluster.producer()
+        ) {
+            var partitionCount = 30;
+            var topics = List.of("topic1", "topic2", "topic3");
+
+            for (var topicName : topics) {
+                cluster.createTopic(topicName, partitionCount, (short) BROKER_COUNT);
+            }
+
+            List<TopicPartition> partitions = new ArrayList<>();
+            for (var topic : topics) {
+                for (var i = 0; i < partitionCount; i++) {
+                    partitions.add(new TopicPartition(topic, i));
+                }
+            }
+
+            assertEquals(0, consumer.assignment().size());
+            consumer.subscribe(topics);
+            awaitAssignment(consumer, Set.copyOf(partitions));
+
+            List<ProducerRecord<byte[], byte[]>> producerRecords = new ArrayList<>();
+            for (var partition : partitions) {
+                producerRecords.addAll(sendRecords(producer, partition, partitionCount, System.currentTimeMillis(), -1));
+            }
+
+            List<ConsumerRecord<byte[], byte[]>> consumerRecords = consumeRecords(consumer, producerRecords.size());
+
+            var expected = new HashSet<>();
+            for (ProducerRecord<byte[], byte[]> record : producerRecords) {
+                expected.add(new RecordInformation(
+                    record.topic(),
+                    record.partition(),
+                    new String(record.key()),
+                    new String(record.value()),
+                    record.timestamp())
+                );
+            }
+
+            var actual = new HashSet<>();
+            for (ConsumerRecord<byte[], byte[]> record : consumerRecords) {
+                actual.add(new RecordInformation(
+                    record.topic(),
+                    record.partition(),
+                    new String(record.key()),
+                    new String(record.value()),
+                    record.timestamp())
+                );
+            }
+
+            assertEquals(expected, actual);
```

Comment:
Could you please consider comparing the fields one by one to produce accurate error message?
```java
            for (var i = 0; i < consumerRecords.size(); i++) {
                var producerRecord = producerRecords.get(i);
                var consumerRecord = consumerRecords.get(i);
                assertEquals(producerRecord.topic(), consumerRecord.topic());
                assertEquals(producerRecord.partition(), consumerRecord.partition());
            }
```

Reply:
Sure, addressed it :)